ql/src/java/org/apache/hadoop/hive/ql/exec/CommonJoinOperator.java

```diff
@@ -523,11 +533,19 @@ private boolean createForwardJoinObject(boolean[] skip) throws HiveException {
         forward = true;
       }
     }
+    return forward;
+  }
+
+  // returns whether a record was forwarded
+  private boolean createForwardJoinObject(boolean[] skip, boolean antiJoin) throws HiveException {
+    boolean forward = fillFwdCache(skip);
     if (forward) {
       if (needsPostEvaluation) {
         forward = !JoinUtil.isFiltered(forwardCache, residualJoinFilters, residualJoinFiltersOIs);
       }
-      if (forward) {
+
+      // For anti join, check all right side and if nothing is matched then only forward.
```

Comment:
Not sure I fully understand the comment here -- !forward (false) and antijoin (true) will still skip the object

Reply:
For anti join we don't emit the record here. It's done after all the records are checked and none of the record matches the condition. Here if forward is false we don't forward and as its a "&" we don't forward for anti join == true even if forward is true.

Reply:
Ok makes sense now -- so maybe we should just mention that for anti-join we dont forward at this point

Reply:
done